File: 백준/Silver/9012. 괄호/괄호.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

input.shift();

input.forEach((string) => {
  const stack = [];
  let result = "YES";

  for (let el of string) {
    if (el === "(") {
      stack.push(el);
    }
    if (el === ")") {
      if (stack.pop() === undefined) {
        result = "NO";
      }
    }
  }

  if (stack.length !== 0) {
    result = "NO";
  }

  console.log(result);
});

File: 백준/Silver/9012. 괄호/괄호.py
T = int(input())

for _ in range(T):
    stack = []
    brackets = list(input())
    for bracket in brackets:
        if not stack or bracket == "(":
            stack.append(bracket)
        elif bracket == ")":
            if stack[-1] == "(":
                stack.pop()
            else:
                break

    if not stack:
        print("YES")
    else:
        print("NO")